2-explicit-polymorphism/2-explicit-polymorphism.md

```diff
@@ -0,0 +1,147 @@
+# Introduction
+
+Currently in Hazel every function has to be assigned a monomorphic type.
+This makes it impossible to write code that operates on more than one type
+of data. We currently have an example that implements `map : (Num -> Num)
+-> [Num] -> [Num]`. By adopting this proposal, Hazel's map function can
+have the type `forall a. (a -> a) -> [a] -> [a]`, and thereby be useful for
+a larger set of values.
+
+Specifically, we propose extending Hazel with two type constructs
+and two expression constructs, bringing Hazel's type system up to
+a System F with holes than a Simply Typed Lambda Calculus with holes.
+
+# Type Language
+
+We'd have to extend UHTyp.t with `TyVar` and `Forall` cases:
+
+```reason
+/* UHTyp.re */
+
+type operand =
+...
+| Forall(TPat.t, t);
+| TyVar(VarErrStatus.t, Var.t)
+  ```
+
+  A type pattern is the part of the language between the `forall` and the `.`
+  in `forall a. <type>`. Type patterns will be very simple to begin with:
+
+  ```reason
+  /* TPat.t */
+  type t =
+  | Hole(MetaVar.t)
+| Var(Var.t)
+  ```
+
```

Comment:
We should also talk about HTyp.t and how we are using a de Bruijn indexed representation, though maybe not until after introducing the user-facing aspects of the proposal